knowledge-base/adr/structure-of-error-object.md

```diff
@@ -0,0 +1,171 @@
+    at ...
+Caused by:
+Error: Request failed with status code 400
+    at createError (XXX/srv/node_modules/axios/lib/core/createError.js:16:15)
+    at settle (XXX/srv/node_modules/axios/lib/core/settle.js:17:12)
+    at IncomingMessage.handleStreamEnd (XXX/srv/node_modules/axios/lib/adapters/http.js:236:11)
+    at IncomingMessage.emit (events.js:326:22)
+    at endReadableNT (_stream_readable.js:1223:12)
+    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+```
+
+but it is totally unstructured and a pure string. 
+The issue raised by the customer asked to have the root cause in a more structured way like:
+
+```typescript
+interface HttpError {
+    name: string;
+    statucCode:number
+    message: string;    
+}
+```
+This ADR decides the way to improve the situation for the customer.
+
+## Options
+
+### Option A: Extending the Error Interface/Object
+
+We could extend the Error object: 
+
+```typescript
+class HttpError extends Error{
+  constructor(...args) {
+    super(...args); 
+    this.name = "HttpError"; 
+  }
+}
+```
+and return a more concrete Error in the `errorWithCause` method.
+There is a `isAxiosError` to detect HTTP errors [see here](https://github.com/axios/axios/pull/1419)
+
+Pros:
+- Typed information specific to error: Http, IO, ...
+- Detailed information like HttpStatus etc can be accessed easily.
+
+Cons:
+- Not very JavaScripty
+- We need to find the right type in the `errorWithCause` -> big exception mapper factory
+- The consumer needs some kind of "down cast" to find the actual error type returned (catch blocks, instanceof check)
+This becomes even more problematic since there are no `throws declaration` possible.
+
+### Option B: Include the Root cause error Message 
+
+We could at least include  a `root` property to the error:
+
+```typescript
+interface ErrorWithRoot {
+    name: string;
+    message: string;
+    stack?: string;
+    rootMessage?:string   
+}
+```
+
+the variable `rootMessage` would contain `Request failed with status code 400` in the example above.
+
+Pros:
+- You can access the root message  quickly
+- No inheritance
+
+Cons:
+- No structured information on the kind of error
+- A check if the property is there is needed
+
+### Option C: Make the RequestBuilder non-throwy
+
+In the API discussion [ADR](https://github.com/SAP/cloud-sdk-js/pull/709) we discussed a change to include more request and response data:
+
+```typescript
+const [buPa, req, res]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().getAll().executeRaw(destination);
+```
+
+One could include a `httpNoThrow` flag to the `execute` methods. 
+If switched on the request builder will not throw HTTP related error but include them in the return. 
+
+```typescript
+const [buPa, httpError]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().httpNoThrow().getAll().execute(destination);
```

Comment:
Maybe adjust the type? `const [buPa, httpError]: [BusinessPartner, Request, Response]`